crates/path_trie/src/lib.rs

```diff
@@ -0,0 +1,967 @@
+                }
+                (false, true) => {
+                    let new = new_winner.unwrap();
+                    from_clobbers.push((cur.clone(), new.clone()));
+                    collect_removed_subtree(n, cur, old_rank, to_clobbers);
+                    // don't descend
+                }
+                (true, false) => {
+                    let old = old_winner.unwrap();
+                    to_clobbers.push((cur.clone(), old.clone()));
+                    collect_new_winners(n, cur, new_rank, from_clobbers);
+                    // don't descend further: new_winners handles it
+                }
+                (false, false) => {
+                    // Recurse normally
+                    for (comp, child) in &n.children {
+                        cur.push(comp);
+                        dfs(child, cur, old_rank, new_rank, to_clobbers, from_clobbers);
+                        cur.pop();
+                    }
+                }
+            }
+        }
+
+        {
+            let sorted_packages = self
+                .packages
+                .iter()
+                .cloned()
+                .sorted()
+                .collect::<Vec<PackageName>>();
+
+            let new_sorted_packages = new_order
+                .iter()
+                .cloned()
+                .sorted()
+                .collect::<Vec<PackageName>>();
+
+            assert_eq!(
+                &sorted_packages, &new_sorted_packages,
+                "Expected just reordering, got something else.
+Old:
+{sorted_packages:#?}
+New:
+{new_sorted_packages:#?}
+"
+            );
+        }
+
+        // Package with highest priority will have biggest number.
+        let old_rank = rank(self.packages.iter().rev());
+        let new_rank = rank(new_order.iter());
+
+        let mut to_clobbers = Vec::new();
+        let mut from_clobbers = Vec::new();
+
+        let mut buf = PathBuf::new();
+        dfs(
+            &self.root,
+            &mut buf,
+            &old_rank,
+            &new_rank,
+            &mut to_clobbers,
+            &mut from_clobbers,
+        );
+
+        self.packages.clear();
+        for pkg in new_order.iter().rev() {
+            self.packages.insert(pkg.clone());
+        }
+
+        (to_clobbers, from_clobbers)
+    }
+
+    /// Move files on-disk:
+    ///
+    /// - For each `(p,pkg)` in `to_clobbers`:  `base/p` → `clobbers/pkg/p` if `base/p` exists and dest doesn’t.
+    /// - For each `(p,pkg)` in `from_clobers`: `clobbers/pkg/p` → `base/p` if source exists and dest doesn’t.
+    pub fn sync_clobbers(
+        target_prefix: &Path,
+        clobbers_dir: &Path,
+        to_clobbers: &[(PathBuf, PackageName)],
+        from_clobbers: &[(PathBuf, PackageName)],
+    ) -> io::Result<()> {
+        fn mv(src: PathBuf, dst: PathBuf) -> io::Result<()> {
+            tracing::trace!("Moving from {} to {}", src.display(), dst.display());
+            if let Some(p) = dst.parent() {
```

Comment:
Is this actually needed? I assume since the file was previously clobbered the target directory already exists.

Reply:
But what if we update package? First every file is removed, then new package is in clobbers. It may be that directories do not exists when we move file from clobbers. And if dir exists, then this call has almost no cost https://doc.rust-lang.org/src/std/fs.rs.html#3139-3164.